### PR TITLE
Feature#3 add devise for admin

### LINKED ISSUE #3: Add devise authentication for administrator
# REQUIREMENTS

Implement authentication for admin with the following tools:

 - Devise
 - Bootstrap_form
 - https://github.com/bootstrap-ruby/bootstrap_form/tree/bootstrap-5
 - Create admin user in db:seeds - http://blog.magmalabs.io/2019/11/25/best-practices-using-rails-seeds.html
 - Flash Messages with Devise and Bootstrap: https://github.com/heartcombo/devise/wiki/How-To:-Integrate-I18n-Flash-Messages-with-Devise-and-Bootstrap

# DESIGN

Image 2021-04-12 at 4 46 46 p m

- Example Bootstrap .
- https://getbootstrap.com/docs/5.0/examples/sign-in/
- Form Documentation: https://getbootstrap.com/docs/5.0/forms/overview/

# Acceptance Criteria

- [ ] Admin can sign-in session.
- [ ] Admin can sign-out session. 
- [ ] Admin can't sign-up.
- [ ] Admin can be created in seed
- [ ] Admin should see visual feedback on sign_in and sign_out

### PR DESCRIPTION
**Pull request by anxhe**
# Description

Add devise authentication for administrator:

 - Implement devise gem and customize to allow minimalistic admin authentication.
 - Implement bootstrap_form gem to simplify views form syntax and styles  for autogenerated  devise form, and in the future can facilitate us other forms. 
 - Create admin with db:seed to initialize database.
 - Create home controller to see index and can facilitate sign_out with a link

See issue #3  for more information.
 
 # Testing

- bundle install
- rails server
- rails db:setup
- navigate to: http://localhost:3000/
- sign in with email: admin@example.com and password: password  
- after a successful sign in, click in sign out  to destroy session. 
- can test with other password or user to see failed login feedback.
  
  ## Screenshots 

Image 2021-04-14 at 2 15 18 p m

Image 2021-04-14 at 2 13 43 p m

### CODE REVIEW

**Review by SpidySamurai — DISMISSED** [review]
Looks great!

File: app/views/home/index.html.erb [hunk 1]
```diff
@@ -0,0 +1 @@
+<h1>Hola Admin 😝</h1>
```
Comment (EduardoGHdez) [inline app/views/home/index.html.erb]: 😆

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -266,7 +266,7 @@
   # config.navigational_formats = ['*/*', :html]
 
   # The default HTTP method used to sign out a resource. Default is :delete.
-  config.sign_out_via = :delete
+  config.sign_out_via = %i[delete get]
```
Comment (EduardoGHdez) [inline config/initializers/devise.rb:269]: hmmmmm

Why is this required?

For the `Sign out` button/link?

You can always enforce links to use `delete` method if required.
I don't see a problem on using `get` requests, but wanted to callout that 😄

File: app/views/home/index.html.erb [hunk 3]
```diff
@@ -1 +1,6 @@
-<h1>Hola Admin 😝</h1>
+<nav class="navbar navbar-light bg-light fixed-top">
+  <div class="container-fluid">
+    <span class="mb-0 h1"><h1 class="navbar-brand">Admin</h1></span>
+    <%= link_to "Sign out →", destroy_admin_session_path, class: "btn btn-outline-danger" %>
```
Comment (EduardoGHdez) [inline app/views/home/index.html.erb]: suggestion based on my previous comment:
```ruby
<%= link_to "Sign out →", destroy_admin_session_path, class: "btn btn-outline-danger", method: :delete %>
```

File: Rakefile [hunk 4]
```diff
@@ -2,31 +2,33 @@
...
+    RubyCritic::RakeTask.new("rubycritic:#{dir}") do |task|
+      task.options = %(
+        --custom-format RubyCriticSmallBadge::Report
+        --minimum-score #{RubyCriticSmallBadge.config.minimum_score}
+        --format html
+        --path tmp/rubycritic/#{dir}
+        --format console
+      )
+      task.paths = FileList["#{dir}/**/*.rb"]
+    end
```
Comment (EduardoGHdez) [inline Rakefile]: 🔥

**Review by EduardoGHdez — DISMISSED** [review]
Congratulations! 💯 

Despite it'd look like a big PR, it was super easy to review because of the progressive commits and descriptive messages you wrote


Thanks thanks thanks! 🔥 


:shipit:

File: Rakefile [hunk 5]
```diff
@@ -2,31 +2,33 @@
...
+
+    RubyCritic::RakeTask.new("rubycritic:#{dir}") do |task|
+      task.options = %(
+        --custom-format RubyCriticSmallBadge::Report
+        --minimum-score #{RubyCriticSmallBadge.config.minimum_score}
+        --format html
+        --path tmp/rubycritic/#{dir}
+        --format console
+      )
+      task.paths = FileList["#{dir}/**/*.rb"]
```
Comment (ArielMD) [inline Rakefile]: nice, i like it

**Review by ArielMD — DISMISSED** [review]
that looks good, good job.

**Review by Arkamis — DISMISSED** [review]
LGTM!, nice example

File: app/views/home/index.html.erb (hunk 3 above)
Comment (anxhe) [reply app/views/home/index.html.erb]: Thanks, changed 👍

**Review by Arkamis — APPROVED** [review]
Looks great

File: config/initializers/devise.rb (hunk 2 above)
Comment (anxhe) [reply config/initializers/devise.rb:269]: Oh no , it was to test sign out via URL (before adding an index page and the link)  but later I forgot to add delete method in the link. Thanks

File: app/views/home/index.html.erb (hunk 3 above)
Comment (anxhe) [reply app/views/home/index.html.erb]: Thanks , changed 👍

File: lib/tasks/code_quality.rake [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    RubyCritic::RakeTask.new("rubycritic:#{dir}") do |task|
+      task.options = %(
+        --custom-format RubyCriticSmallBadge::Report
+        --minimum-score #{RubyCriticSmallBadge.config.minimum_score}
+        --format html
+        --path tmp/rubycritic/#{dir}
+        --format console
+      )
```
Comment (MiguelOcegueraM) [inline lib/tasks/code_quality.rake:20]: nice!

File: app/views/devise/sessions/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+          </div>
+        <% end %>
+        <div class="actions d-flex justify-content-center mt-4">
+          <%= f.submit "Log in" , class: 'btn btn-primary col-lg-3' %>
+        </div>
+      <% end %>
+      <%= render "devise/shared/links" %>
+    </div>
+  </div>
+</div>
```
Comment (MiguelOcegueraM) [inline app/views/devise/sessions/new.html.erb:27]: Try to improve your HTML using different types of tags and fewer divs, for a semantic HTML.

**Review by MiguelOcegueraM — COMMENTED** [review]
A very organized PR the sequence of commits is excellent

File: app/views/devise/sessions/new.html.erb (hunk 7 above)
Comment (MiguelOcegueraM) [reply app/views/devise/sessions/new.html.erb:27]: https://dev.to/kenbellows/stop-using-so-many-divs-an-intro-to-semantic-html-3i9i

Comment (SpidySamurai) [reply app/views/devise/sessions/new.html.erb:27]: Yes i recomend to use semantic HTML ! to improve sintaxis

**Review by SpidySamurai — APPROVED** [review]
